addons/website_event/models/event_event.py

```diff
@@ -116,28 +141,66 @@ def _compute_is_participating(self):
             registration;
         """
         current_visitor = self.env['website.visitor']._get_visitor_from_request()
-        base_domain = [('event_id', 'in', self.ids), ('state', 'in', ['open', 'done'])]
+        events_in_self = [('event_id', 'in', self.ids)] if self else []
         if self.env.user._is_public() and not current_visitor:
-            events = self.env['event.event']
-        elif self.env.user._is_public():
-            events = self.env['event.registration'].sudo().search(
-                expression.AND([base_domain, [('visitor_id', '=', current_visitor.id)]])
-            ).event_id
+            return self.env['event.event']
+        if self.env.user._is_public():
+            domain = expression.AND([[
+                ('state', '!=', 'cancel'),
+                ('visitor_id', '=', current_visitor.id),
+            ], events_in_self])
         else:
             if current_visitor:
-                domain = [
+                partner_domain = [
                     '|',
                     ('partner_id', '=', self.env.user.partner_id.id),
                     ('visitor_id', '=', current_visitor.id)
                 ]
             else:
-                domain = [('partner_id', '=', self.env.user.partner_id.id)]
-            events = self.env['event.registration'].sudo().search(
-                expression.AND([base_domain, domain])
-            ).event_id
+                partner_domain = [('partner_id', '=', self.env.user.partner_id.id)]
+            domain = expression.AND([
+                partner_domain,
+                expression.AND([[('state', '!=', 'cancel')], events_in_self])
+            ])
+
+        return self.env['event.registration'].sudo().search(domain).event_id
```

Comment:
Seems a little harsh performance-wise to fetch all the registrations and then the matching event.
Even though it's based on a logged user so there should not be that many.

How about a read_group, grouped on event_id?